# Principal Component Analysis

Why PCA?

  . Visualisation

  . Dimensionality Reduction

  . Multicollinearity
  
  . Noise reduction in the dataset

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

In [3]:
df=pd.read_csv('/content/Iris (1).csv')
df.head()

Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm      Species
0   1            5.1           3.5            1.4           0.2  Iris-setosa
1   2            4.9           3.0            1.4           0.2  Iris-setosa
2   3            4.7           3.2            1.3           0.2  Iris-setosa
3   4            4.6           3.1            1.5           0.2  Iris-setosa
4   5            5.0           3.6            1.4           0.2  Iris-setosa

In [4]:
labels=df["Species"]
data=df.drop(["Id","Species"],axis=1)

In [5]:
print(data.shape)
print(labels.shape)

(150, 4)
(150,)


In [6]:
data.head()

SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm
0            5.1           3.5            1.4           0.2
1            4.9           3.0            1.4           0.2
2            4.7           3.2            1.3           0.2
3            4.6           3.1            1.5           0.2
4            5.0           3.6            1.4           0.2

In [9]:
from sklearn.preprocessing import StandardScaler
standardize_data=StandardScaler().fit_transform(data)
print(standardize_data.shape)

(150, 4)


In [10]:
standardize_data[:5]

array([[-0.90068117,  1.03205722, -1.3412724 , -1.31297673],
       [-1.14301691, -0.1249576 , -1.3412724 , -1.31297673],
       [-1.38535265,  0.33784833, -1.39813811, -1.31297673],
       [-1.50652052,  0.10644536, -1.2844067 , -1.31297673],
       [-1.02184904,  1.26346019, -1.3412724 , -1.31297673]])

# PCA using SCIKIT-LEARN

In [12]:
from sklearn.decomposition import PCA
pca=PCA()
pca.fit(standardize_data)

PCA()

In [13]:
pca.components_

array([[ 0.52237162, -0.26335492,  0.58125401,  0.56561105],
       [ 0.37231836,  0.92555649,  0.02109478,  0.06541577],
       [-0.72101681,  0.24203288,  0.14089226,  0.6338014 ],
       [-0.26199559,  0.12413481,  0.80115427, -0.52354627]])

In [14]:
pca.explained_variance_ratio_

array([0.72770452, 0.23030523, 0.03683832, 0.00515193])

In [17]:
np.cumsum(pca.explained_variance_ratio_)

array([0.72770452, 0.95800975, 0.99484807, 1.        ])

In [19]:
np.arange(len(np.cumsum(pca.explained_variance_ratio_)))

array([0, 1, 2, 3])

**OBSERVATION:** As we can see that only 2 PCA components are explaining 95.8% variance, so lets go ahead and consider these two components.

In [22]:
# configuring the parameteres
# number of components = 2
pca_new=PCA(n_components=2)
pca_new_data=pca_new.fit_transform(standardize_data)

print('shape=',pca_new_data.shape)

shape= (150, 2)


In [23]:
pca_df=pd.DataFrame(data=pca_new_data,columns=['1st_principal','2nd_principal'])
pca_df['Labels']=labels.values
pca_df.head()

1st_principal  2nd_principal       Labels
0      -2.264542       0.505704  Iris-setosa
1      -2.086426      -0.655405  Iris-setosa
2      -2.367950      -0.318477  Iris-setosa
3      -2.304197      -0.575368  Iris-setosa
4      -2.388777       0.674767  Iris-setosa